### PR TITLE
Assignment 6

### PR DESCRIPTION
**Pull request by GamerBah**
Still not sure whether I'm doing the code analysis stuff right since I'm using Rider.

### CODE REVIEW

**Comment by dbosco051394** [comment]
It's looking good to me. Nice work!

As for the analyzers, I couldn't find anything that suggest Rider can use the analyzers used in Visual Studio. I think the analyzers are Visual Studio only.

**Comment by GamerBah** [comment]
> As for the analyzers, I couldn't find anything that suggest Rider can use the analyzers used in Visual Studio. I think the analyzers are Visual Studio only.

Well I know I can do inspections and disable the warnings using a ReSharper comment, but since we're not supposed to disable them, the only other option in Rider is to locally change the warning level of the inspection for the project. @Keboo is this the same as what is done in Visual Studio?

**Comment by dbosco051394** [comment]
As far as I understood the requirements on this assignment, we needed to add the specific analyzers to our project that were listed and comply with them to our best ability. I'm not too familiar with Rider to be honest, so I think talking to the instructors on this one either in class or on the chat room would be wise.

File: Assignment6.Tests/Assignment6.Tests.csproj [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        <PackageReference Include="coverlet.collector" Version="1.0.1" />
+    </ItemGroup>
+
+    <ItemGroup>
+        <ProjectReference Include="..\Assignment6\Assignment6.csproj" />
+    </ItemGroup>
+
+    <ItemGroup>
+      <ProjectReference Include="..\Assignment6\Assignment6.csproj" />
+    </ItemGroup>
```
Comment (ashermancinelli) [inline Assignment6.Tests/Assignment6.Tests.csproj]: double reference could be deleted

File: Assignment6/Array.cs [hunk 2]
```diff
@@ -1,10 +1,129 @@
...
+using System.Data;
+using System.Linq;
+
+namespace Assignment6
+{
+
+    public class ArrayCollection<T> : ICollection<T>
+    {
+
+        private T[] _array;
```
Comment (ashermancinelli) [inline Assignment6/Array.cs]: I think this is supposed to be `_Array`

**Comment by MarkMichaelis** [comment]
##  (GamerBah)

## Assignment6

| Build/Compile (Build Succeeded: 0 Error(s), 25 Warning(s))| **10** |
| --------------------- | ------ |
| Assignment compiles even though major pieces might be missing, or there are significant problems: **✔** |  **5** |
| The assignment contains only a few required elements or has significant problems. | **5** |


| Assignment  | **20** |
| --------------------- | ------ |
| Supports collecting a fixed number of items once instantiated. | ✔ |
| Has a constructor via which you can specify fixed width.  (Other constructors are welcome.) | ✔ |
| Handle null appropriately when items are missing (this is likely a judgement call so justify your answer in the comments.) | ✔ |
| Throw an exception if an item doesn't exist. | ✔ |
| Include an index operator. | ✔ |
| Support using foreach over the items in the `Array<T>` class. | ✔ |
| Support a collection initializer. | ✔ |
| Implement `System.Collections.Generic.ICollection<T>`. | ✔ |
| Ignore support for multi-dimensional arrays. | ✔ |
| Add a `Capacity` property to return the capacity (the value specified in the constructor) of the collection. | ✔ |


| Project Analysis |   |
| -----------------| - |
| Nullable Enabled: | **✔** |
| CodeReview Performed | **✔** |
| Add the code analysis packages and appropriately handle all warnings* |  **✔** |
*<sub><sup>`IntelliTectAnalyzer.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.CodeQuality.Analyzers.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.NetFramework.Analyzers.dll`. (It is okay to appropriately disable rules such as  localization/culture rules.)</sub></sup>


| ProjectName | Nullable | Solutions | Analyzers | Disabled Warnings | Waning Level |
| -- | -- | -- | -- | -- | -- |
| Assignment6 | Enable | Assignment6.sln |  | CS1701, CS1702, CS2008 | 4 |
| Assignment6.Tests | Enable | Assignment6.sln |  | CS1701, CS1702 | 4 |


| Testing | **15** |
| --------------------- | ------ |
| Test Run Succeeded: Total tests=19, Passed=19, Failed=0, Skipped=0 | **84%** |
| Unit test project fails to compile, most test fails, or contains significant problems. | **10** |
| Most assigned work is tested or contain some errors. | **5** |
| All assigned work is tested | **0** |


| Extra Credit | **0** |
| --------------------- | ------ |
| Write a cast operator to a C# array or a `System.Collections.Generic.List<T>`. | ✔ |
| Add support for an `IReadOnlyArray` interface (for you to define) that provides readonly access to the count and each element in the array.  Be sure to appropriately decorate it for covariance and contravariance | ✔ |

**Pull request closed without merging**